### PR TITLE
sd_ass: introduce sub-ass-prune-delay

### PR DESCRIPTION
**Pull request by llyyr**

### CODE REVIEW

**Comment by sfan5** [comment]
do we really have to make this configurable?
if you seek events will be re-read anyway so something like 10s should be guaranteed safe

**Comment by kasper93** [comment]
Why is this even configurable on libass side? Why can't they manage their internal memory cache?

**Comment by llyyr** [comment]
> Why is this even configurable on libass side? Why can't they manage their internal memory cache?

This is a new option, the use case is playback of a video file made solely out of ASS vector-art. libass currently keeps all events in memory until `ass_flush_events` or `ass_free_track` is called.

> do we really have to make this configurable?

Setting it to a really small number should be fine too

**Comment by llyyr** [comment]
Changed it to just unconditionally enable pruning if the API is available. I tested it briefly and nothing seems broken by it

**Comment by cubicibo** [comment]
> Why is this even configurable on libass side? Why can't they manage their internal memory cache?

libass cannot know the user intention and use-case.
- User may parse an entire script once then perform random seek forward and backward: libass must not do anything.
- User may only render content incrementally: libass can optionally discard expired events.
- User can playback content where subtitle and video packets are multiplexed. Relying on libass cache can be faster than flushing everything and re-parsing.
- Continuous stream with closed captions (e.g. Twitch) no longer result in memory leaks.

This should probably be enabled by default with a reasonable time window for ASS tracks multiplexed with AV content, like MKVs.

**Comment by sfan5** [comment]
> I tested it briefly and nothing seems broken by it

Wouldn't it also break sub-seek if the previous event is older than prune-delay?

**Comment by llyyr** [comment]
> Wouldn't it also break sub-seek if the previous event is older than prune-delay?

Yes, so sub-seek and switching to play-dir=- during runtime is broken if we enable it unconditionally

**Comment by Samillion** [comment]
This would solve https://github.com/mpv-player/mpv/issues/9014 right?

**Comment by Samillion** [comment]
If the concern about seconds being too fast or would unnecessarily prune cache for it to re-render once again, would it be viable to add a "file" option. As in, cache stays as long as the current file/stream is running, on next file load, prunes cache and re-render.

This would at least solve a scenario for those that leave mpv open for long times, like the museum guy the other day or the same issue I linked above.

**Comment by llyyr** [comment]
> This would solve #9014 right?

No, this will only apply to the subtitle track, not the OSD. I don't believe this is necessary or even warranted for the OSD since as avih said in that issue: `Sure, with OSD/OSC it uses more ram (and grows quicker initially), because libass has a big cache. However, that should eventually stop growing`

It shouldn't be growing infinitely in the first place, so if it does that's a whole another issue.

**Comment by Samillion** [comment]
Noted. Thank you for the response.

**Comment by kasper93** [comment]
Have anyone tested it?

**Comment by llyyr** [comment]
> Have anyone tested it?

Yes, I tested it. Works as expected, I listed the things it breaks in the docs that's why I don't think we can enable it by default unfortunately

File: DOCS/man/options.rst [hunk 1]
```diff
@@ -2533,6 +2533,22 @@ Subtitles
     ``complex`` is the default. If libass hasn't been compiled against HarfBuzz,
     libass silently reverts to ``simple``.
 
+``--sub-ass-prune-delay=<-1|seconds>``
```
Comment (kasper93) [inline DOCS/man/options.rst:2536]: This needs interface changes entry.

Comment (llyyr) [reply DOCS/man/options.rst:2536]: Done but we should remove this line https://github.com/mpv-player/mpv/blob/master/DOCS/interface-changes.rst?plain=1#L20-L21

File: DOCS/interface-changes/sub-ass-prune-delay.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+add `sub-ass-prune-delay` option to control libass automatica pruning of past
```
Comment (kasper93) [inline DOCS/interface-changes/sub-ass-prune-delay.txt:1]: typo

File: DOCS/man/options.rst (hunk 1 above)
Comment (kasper93) [reply DOCS/man/options.rst:2536]: I think we list each new option for a long time.

**Review by kasper93 — APPROVED** [review]

**Pull request merged**